packages/expo-updates/ios/EXUpdates/AppController.swift

```diff
@@ -38,12 +39,32 @@ public class AppController: NSObject, AppLoaderTaskDelegate, ErrorRecoveryDelega
   private static let ErrorDomain = "EXUpdatesAppController"
   private static let EXUpdatesEventName = "Expo.nativeUpdatesEvent"
 
+  // Events for the legacy UpdateEvent JS listener
   public static let UpdateAvailableEventName = "updateAvailable"
   public static let NoUpdateAvailableEventName = "noUpdateAvailable"
   public static let ErrorEventName = "error"
 
+  // Events that will be used to track state changes
```

Comment:
What do you envision the state looking like? I had imagined it as being something like:
```
{  
  isCheckingForNewUpdate: bool;  
  newUpdate: Update | null; // next update that will be launched when restarted  
  newUpdateError: Error | null; // error that may have occurred  
  isNewUpdateDownloaded: bool; // whether the new update is downloaded, maybe this should be a property within newUpdate  
  newUpdateDownloadProgress: number | null; // maybe combine with `isNewUpdateDownloaded` and just have 100 mean downloaded
}
```

I think it may be worth using a state machine to keep track of the state. But that is the next PR I guess.

The reason I bring this up here in the review is that I think a lot of these events may break the abstraction direction a bit (AppController shouldn't know about the Module). I had imagined this more being purely a delegation pattern. Concretely, this means that all events would be emitted by the classes underneath this (by either delegate calls or nsnotificationcenter if necessary) and that this class would accumulate them. This may require some changes to the module though, so I'm not sure of the scope of this.

Just for example sake, here's how I imagined `fetchUpdateAsync` would work:
1. module calls `RemoteAppLoader.loadUpdate`
2. That remote app loader has a reference to AppController somehow (not sure if this is true? should we make it true if not? or does that break some abstraction in the interface? what even is the interface lol?)
3. The remote app loader does it's stuff and propagates events to its reference to AppController through its reference.

Reply:
The state machine is now implemented in this PR.